File: cplusplus/src/message.c
#include <string.h>

#include "autograph.h"
#include "constants.h"
#include "external.h"

size_t calculate_padded_size(const size_t plaintext_size) {
  return plaintext_size + PADDING_BLOCK_SIZE -
         (plaintext_size % PADDING_BLOCK_SIZE);
}

size_t autograph_ciphertext_size(const size_t plaintext_size) {
  return calculate_padded_size(plaintext_size) + TAG_SIZE;
}

size_t autograph_plaintext_size(const size_t ciphertext_size) {
  return ciphertext_size - TAG_SIZE;
}

void pad(uint8_t *padded, const size_t padded_size, const uint8_t *plaintext,
         const size_t plaintext_size) {
  zeroize(padded, padded_size);
  memmove(padded, plaintext, plaintext_size);
  padded[plaintext_size] = PADDING_BYTE;
}

bool encrypt_plaintext(uint8_t *ciphertext, const uint8_t *key,
                       const uint8_t *nonce, const uint8_t *plaintext,
                       const size_t plaintext_size) {
  size_t padded_size = calculate_padded_size(plaintext_size);
  uint8_t padded[padded_size];
  pad(padded, padded_size, plaintext, plaintext_size);
  return encrypt(ciphertext, key, nonce, padded, padded_size);
}

bool increment_nonce(uint8_t *nonce) {
  size_t offset = NONCE_SIZE - 4;
  uint32_t index = get_uint32(nonce, offset);
  if (index == UINT32_MAX) {
    return false;
  }
  set_uint32(nonce, offset, index + 1);
  return true;
}

bool encrypt_fail(uint8_t *key, uint8_t *nonce) {
  zeroize(key, SECRET_KEY_SIZE);
  zeroize(nonce, NONCE_SIZE);
  return false;
}

bool autograph_encrypt(uint32_t *index, uint8_t *ciphertext, uint8_t *key,
                       uint8_t *nonce, const uint8_t *plaintext,
                       const size_t plaintext_size) {
  if (!increment_nonce(nonce)) {
    return encrypt_fail(key, nonce);
  }
  *index = get_uint32(nonce, NONCE_SIZE - 4);
  if (!encrypt_plaintext(ciphertext, key, nonce, plaintext, plaintext_size)) {
    return encrypt_fail(key, nonce);
  }
  return true;
}

size_t calculate_unpadded_size(const uint8_t *padded,
                               const size_t padded_size) {
  if (padded_size == 0 || (padded_size % PADDING_BLOCK_SIZE) > 0) {
    return 0;
  }
  for (size_t i = padded_size - 1; i >= (padded_size - PADDING_BLOCK_SIZE);
       --i) {
    uint8_t byte = padded[i];
    if (byte == PADDING_BYTE) {
      return i;
    }
    if (byte != 0) {
      return 0;
    }
  }
  return 0;
}

bool unpad(size_t *unpadded_size, const uint8_t *padded,
           const size_t padded_size) {
  size_t size = calculate_unpadded_size(padded, padded_size);
  if (size == 0) {
    return false;
  }
  *unpadded_size = size;
  return true;
}

bool decrypt_ciphertext(uint8_t *plaintext, size_t *plaintext_size,
                        const uint8_t *key, const uint8_t *nonce,
                        const uint8_t *ciphertext,
                        const size_t ciphertext_size) {
  if (decrypt(plaintext, key, nonce, ciphertext, ciphertext_size)) {
    return unpad(plaintext_size, plaintext,
                 autograph_plaintext_size(ciphertext_size));
  }
  return false;
}

size_t get_skipped_index_offset(const uint8_t *skipped_indexes,
                                const size_t skipped_indexes_size) {
  size_t max_offset = skipped_indexes_size - 4;
  if (get_uint32(skipped_indexes, max_offset) > 0) {
    return skipped_indexes_size;
  }
  size_t offset = 0;
  while (offset <= max_offset) {
    if (get_uint32(skipped_indexes, offset) == 0) {
      return offset;
    }
    offset += 4;
  }
  return skipped_indexes_size;
}

bool skip_index(const uint8_t *nonce, uint8_t *skipped_indexes,
                const size_t skipped_indexes_size) {
  size_t offset =
      get_skipped_index_offset(skipped_indexes, skipped_indexes_size);
  if (offset == skipped_indexes_size) {
    return false;
  }
  memmove(skipped_indexes + offset, nonce + NONCE_SIZE - 4, 4);
  return true;
}

size_t get_skipped_index(uint32_t *index, uint8_t *nonce,
                         const uint8_t *skipped_indexes,
                         const size_t skipped_indexes_size,
                         const size_t offset) {
  if (offset == 0) {
    zeroize(nonce, NONCE_SIZE);
  }
  if (offset == skipped_indexes_size) {
    return 0;
  }
  *index = get_uint32(skipped_indexes, offset);
  memmove(nonce + NONCE_SIZE - 4, skipped_indexes + offset, 4);
  return offset + 4;
}

void delete_skipped_index(uint8_t *skipped_indexes,
                          const size_t skipped_indexes_size,
                          const size_t next_offset) {
  size_t offset = next_offset - 4;
  size_t last_offset = skipped_indexes_size - 4;
  if (offset != last_offset) {
    memmove(skipped_indexes + offset, skipped_indexes + last_offset, 4);
  }
  zeroize(skipped_indexes + last_offset, skipped_indexes_size - last_offset);
}

bool decrypt_skipped(uint32_t *index, uint8_t *plaintext,
                     size_t *plaintext_size, const uint8_t *key,
                     uint8_t *skipped_indexes,
                     const size_t skipped_indexes_size,
                     const uint8_t *ciphertext, const size_t ciphertext_size) {
  uint8_t nonce[NONCE_SIZE];
  size_t offset =
      get_skipped_index(index, nonce, skipped_indexes, skipped_indexes_size, 0);
  while (offset < skipped_indexes_size) {
    if (decrypt_ciphertext(plaintext, plaintext_size, key, nonce, ciphertext,
                           ciphertext_size)) {
      delete_skipped_index(skipped_indexes, skipped_indexes_size, offset);
      return true;
    }
    offset = get_skipped_index(index, nonce, skipped_indexes,
                               skipped_indexes_size, offset);
  }
  return false;
}

bool decrypt_fail(uint8_t *key, uint8_t *nonce, uint8_t *skipped_indexes,
                  const size_t skipped_indexes_size) {
  zeroize(key, SECRET_KEY_SIZE);
  zeroize(nonce, NONCE_SIZE);
  zeroize(skipped_indexes, skipped_indexes_size);
  return false;
}

bool autograph_decrypt(uint32_t *index, uint8_t *plaintext,
                       size_t *plaintext_size, uint8_t *key, uint8_t *nonce,
                       uint8_t *skipped_indexes,
                       const size_t skipped_indexes_size,
                       const uint8_t *ciphertext,
                       const size_t ciphertext_size) {
  bool success =
      decrypt_skipped(index, plaintext, plaintext_size, key, skipped_indexes,
                      skipped_indexes_size, ciphertext, ciphertext_size);
  while (!success) {
    if (!increment_nonce(nonce)) {
      return decrypt_fail(key, nonce, skipped_indexes, skipped_indexes_size);
    }
    *index = get_uint32(nonce, NONCE_SIZE - 4);
    success = decrypt_ciphertext(plaintext, plaintext_size, key, nonce,
                                 ciphertext, ciphertext_size);
    if (!skip_index(nonce, skipped_indexes, skipped_indexes_size)) {
      return decrypt_fail(key, nonce, skipped_indexes, skipped_indexes_size);
    }
  }
  return true;
}

File: go/message/message.go
package message

import (
	"fmt"
	"math"

	c "github.com/christoffercarlsson/autograph/go/constants"
	e "github.com/christoffercarlsson/autograph/go/external"
	t "github.com/christoffercarlsson/autograph/go/types"
)

func CalculatePaddedSize(plainText *[]byte) int {
	size := len(*plainText)
	return size + int(c.PADDING_BLOCK_SIZE) - (size % int(c.PADDING_BLOCK_SIZE))
}

func Pad(plainText *[]byte) t.Bytes {
	bytes := (*plainText)[:]
	paddedSize := CalculatePaddedSize(plainText)
	for range paddedSize - len(bytes) {
		bytes = append(bytes, 0)
	}
	bytes[len(*plainText)] = c.PADDING_BYTE
	return bytes
}

func CalculateUnpaddedSize(padded *[]byte) int {
	size := len(*padded)
	if size == 0 || (size%int(c.PADDING_BLOCK_SIZE)) > 0 {
		return 0
	}
	for i := size - 1; i >= (size - int(c.PADDING_BLOCK_SIZE)); i -= 1 {
		byte := (*padded)[i]
		if byte == c.PADDING_BYTE {
			return i
		}
		if byte != 0 {
			return 0
		}
	}
	return 0
}

func UnPad(plainText *[]byte) error {
	size := CalculateUnpaddedSize(plainText)
	if size == 0 {
		return fmt.Errorf("unpad calculate unpadded size error")
	}
	// TODO is this an error?
	*plainText = (*plainText)[:size]
	return nil
}

func getIndex(nonce *t.Nonce) uint32 {
	nonceSlice := nonce[:]
	return e.GetUint32(&nonceSlice, int(c.NONCE_SIZE-4))
}

func setIndex(nonce *t.Nonce, index uint32) {
	nonceSlice := nonce[:]
	e.SetUint32(&nonceSlice, int(c.NONCE_SIZE-4), index)
}

func IncrementNonce(nonce *t.Nonce, err error) error {
	index := getIndex(nonce)
	if index == math.MaxUint32 {
		return fmt.Errorf("nonce index out of bounds error")
	}
	setIndex(nonce, index+1)
	return nil
}

func CreateCiphertext(plainText *[]byte) t.Bytes {
	return make([]byte, CalculatePaddedSize(plainText)+int(c.TAG_SIZE))
}

func Encrypt(
	key *t.SecretKey,
	nonce *t.Nonce,
	plainText *[]byte,
) (uint32, []byte, error) {
	incrementNonceError := fmt.Errorf("increment nonce error")
	err := IncrementNonce(nonce, incrementNonceError)
	if err != nil {
		return 0, []byte{}, err
	}
	cipherText := CreateCiphertext(plainText)
	padded := Pad(plainText)
	success := e.Encrypt(&cipherText, key, nonce, &padded)
	if !success {
		return 0, []byte{}, fmt.Errorf("encrypt error")
	}
	return getIndex(nonce), cipherText, nil
}

func CreatePlaintext(cipherText *[]byte) t.Bytes {
	return make([]byte, len(*cipherText)+int(c.TAG_SIZE))
}

func DecryptCipherText(
	key *t.SecretKey,
	nonce *t.Nonce,
	cipherText *[]byte,
) (uint32, []byte, error) {
	plainText := CreatePlaintext(cipherText)
	success := e.Decrypt(&plainText, key, nonce, cipherText)
	if !success {
		return 0, []byte{}, fmt.Errorf("decrypt error")
	}
	err := UnPad(&plainText)
	if err != nil {
		return 0, []byte{}, err
	}
	return getIndex(nonce), plainText, nil
}

func DecryptSkipped(
	key *t.SecretKey,
	skippedIndexes *[]uint32,
	cipherText *[]byte,
) (uint32, []byte, error) {
	nonce := [c.NONCE_SIZE]byte{}
	for i, skippedIndex := range *skippedIndexes {
		if skippedIndex == 0 {
			continue
		}
		setIndex(&nonce, skippedIndex)
		index, plainText, err := DecryptCipherText(key, &nonce, cipherText)
		if err != nil {
			continue
		}
		(*skippedIndexes)[i] = 0
		return index, plainText, nil
	}
	return 0, []byte{}, fmt.Errorf("decrypt skipped fail")
}

func skipIndex(skippedIndexes *[]uint32, nonce *t.Nonce) error {
	index := getIndex(nonce)
	for i, skippedIndex := range *skippedIndexes {
		if skippedIndex == 0 {
			(*skippedIndexes)[i] = index
			return nil
		}
	}
	return fmt.Errorf("skip index failure")
}

func Decrypt(
	key *t.SecretKey,
	nonce *t.Nonce,
	skippedIndexes *[]uint32,
	cipherText *[]byte,
) (uint32, []byte, error) {
	i, plainText, err := DecryptSkipped(key, skippedIndexes, cipherText)
	if err == nil {
		return i, plainText, nil
	}
	for {
		err := IncrementNonce(nonce, fmt.Errorf("increment nonce error"))
		if err != nil {
			return 0, []byte{}, err
		}
		i, plainText, err = DecryptCipherText(key, nonce, cipherText)
		if err != nil {
			err := skipIndex(skippedIndexes, nonce)
			if err != nil {
				return 0, []byte{}, err
			}
		} else {
			return i, plainText, nil
		}
	}
}
